SPEC.md

```diff
@@ -201,25 +201,161 @@ Note that logging mechanisms other than stdout and stderr are not required by th
 Additional isolators will be added to this specification over time.
 
-|Name|Type|Schema|Example|
-|---------------------------|------|------------------------------------|------------------------------------|
-|cpu/shares                 |string|"&lt;uint&gt;"                      |"4096"                              |
-|memory/limit               |string|"&lt;bytes&gt;"                     |"1G", "5T", "4K"                    |
-|cpu/mask                   |string|"&lt;cpu range&gt;"                 |"0", "0-3", "0,2,4-7"               |
-|block-io/read-bandwidth    |string|"&lt;path to file&gt; &lt;bytes&gt;"|"/tmp 1K"                           |
-|block-io/write-bandwidth   |string|"&lt;path to file&gt; &lt;bytes&gt;"|"/tmp 1K"                           |
-|network-io/read-bandwidth  |string|"&lt;device name&gt; &lt;bytes&gt;" |"eth0 100M"                         |
-|network-io/write-bandwidth |string|"&lt;device name&gt; &lt;bytes&gt;" |"eth0 100M"                         |
-|capabilities/bounding-set  |string|"&lt;cap&gt; &lt;cap&gt; ..."       |"CAP_NET_BIND_SERVICE CAP_SYS_ADMIN"|
+An isolator is a standalone JSON object with only one required field: "name".
+All other fields are specific to the isolator.
 
-#### Types
+An executor MAY ignore isolators that it does not understand and run the container without them.
+But, an executor MUST make information about which isolators were ignored, enforced or modified to the user.
+An executor MAY implement a "strict mode" where an image cannot run unless all isolators are in place.
 
-* uint: base 10 formatted unsigned int as a string
-* bytes: Suffix to a base 10 int to make it a K, M, G, or T for base 1024
+### Linux Isolators
+
+These isolators are specific to the Linux kernel and are impossible to represent as a 1-to-1 mapping on other kernels.
+The first example is "capabilities" but this will be expanded to include things such as SELinux, SMACK or AppArmor.
+
+#### linux/capabilities-remove-set
+
+* Scope: app
+
+**Parameters:**
+* **set** list of capabilities that will be removed from the process's capabilities bounding set
+
+```
+"name": "linux/capabilities-remove-set",
+"set": [
+	"CAP_SYS_PTRACE",
+]
+```
+
+#### linux/capabilities-retain-set
+
+* Scope: app
+
+**Parameters:**
+* **set** list of capabilities that will be retained in the process's capabilities bounding set
+
+```
+"name": "linux/capabilities-retain-set",
+"set": [
+	"CAP_KILL",
+	"CAP_CHOWN"
+]
+```
+
+### Resource Isolators
+
+#### resource/cpu
+
+* Scope: app/container
+
+**Parameters:**
+* **request** milli-cores that are requested
+* **limit** milli-cores that can be consumed before the Kernel temporarily throttles the process
+
+```
+"name": "resource/cpu",
+"request": "250",
+"limit": "500"
+```
+
+**Note**: a milli-core is the milli-seconds/second that the app will be able to run. e.g. 1000 would represent full use of a single CPU core every second.
```

Comment:
Millicores are a concrete unit.  Shares are a proportional unit.  Guess
which one makes sense to users and which one makes sense to kernel hackers.
On Feb 24, 2015 8:06 AM, "Alban" <notifications@github.com> wrote:

> In SPEC.md <https://github.com/appc/spec/pull/192#discussion_r25263968>:
>
> > +
> > +#### resource/cpu
> > +
> > +* Scope: app/container
> > +
> > +**Parameters:**
> > +* **request** milli-cores that are requested
> > +* **limit** milli-cores that can be consumed before the Kernel temporarily throttles the process
> > +
> > +```
> > +"name": "resource/cpu",
> > +"request": "250",
> > +"limit": "500"
> > +```
> > +
> > +**Note**: a milli-core is the milli-seconds/second that the app will be able to run. e.g. 1000 would represent full use of a single CPU core every second.
>
> Where does the name "milli-core" come from?
> It seems close to the CPUShares option in systemd:
>
> http://www.freedesktop.org/software/systemd/man/systemd.cgroup.html#CPUShares=weight
> And it has a default of 1024 for a single CPU core.
>
> https://www.kernel.org/doc/Documentation/scheduler/sched-design-CFS.txt
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/appc/spec/pull/192/files#r25263968>.
>